rbac/management/group/definer.py

```diff
@@ -112,22 +118,22 @@ def clone_default_group_in_public_schema(group, tenant) -> Optional[Group]:
     tenant_default_policy.tenant = tenant
     if Group.objects.filter(name=group.name, platform_default=group.platform_default, tenant=tenant):
         # TODO: returning none can break other code
-        return None
+        return None, relationships
     public_default_roles = Role.objects.filter(platform_default=True, tenant=public_tenant)
 
     group.save()
     tenant_default_policy.group = group
     tenant_default_policy.save()
     tenant_default_policy.roles.set(public_default_roles)
-    return group
+    return group, relationships
 
 
 @transaction.atomic
-def add_roles(group, roles_or_role_ids, tenant, user=None):
+def add_roles(group, roles_or_role_ids, tenant, user=None, relations=None):
```

Comment:
Main thing I am wondering about is if we can avoid passing `relations` here all the way through.

It is a bit awkward right now. At the very least I would rename the parameter. But maybe we can have this return relations, and then have an outer method consolidate them into a single replication event. But, will take a look more tomorrow.

Sorry for the short comment – need to run now and just wanted to jot this down for you to think on :-). Catch up more tomorrow :-).

Reply:
> Main thing I am wondering about is if we can avoid passing relations here all the way through.

Yes definitely I agree! I wanted to know whether changes are ok regard to algorithm and relation generation. Anyway I will rework to avoid passing `relations`.

Reply:
So as I was looking at this I think I have a suggestion for this, which I think would also make the logic easier to understand/validate.

First, it involves using two outbox events. One event will separate the default access, and replicate that default access with new role bindings. The other will update those role bindings to add or remove roles.

I do not think it is worth the complexity to combine those outbox events. Using two separate events does two things: (1) the additional access or removed access will be "more" eventually consistent with RBAC, by one additional event, and (2) it adds potentially redundant work. As for (1), it is already eventually consistent. Because both events are committed in one transaction, they will still be correctly ordered with respect to concurrent changes (e.g. adding/removing users from the group which is itself eventually consistent for default groups anyway). So, it actually doesn't change the semantics in any way to combine those events. To any perspective, the result is the same: access is the default, and then it changes at some point in the future, regardless of whether that happens with one replication event or two.

As for (2) (performance)... I'm just thinking it's premature optimization to think about that much otherwise. The worst case scenario is that you remove all access that was just added. So you'd start by adding all role bindings, and then you'd remove them. Realistically, if you had default access for 30 roles, that means adding ~90 tuples and then removing them. In the grand scheme of things, probably not a big deal. And if it's a problem, we can always improve it later.

So given that, here is what I would suggest:

1. Change the `RelationApiDualWriteGroupHandler` but not _quite_ as much as in this PR currently. Maybe just keep the separate replicate piece or let it take multiple roles at once. So change `replicate_added_role` to accept multiple roles, and then we just add them all together in one method call. Or separate `add_role` from `replicate_added_role` and then keep the separate `replicate()` method. (but we don't need the custom group param or other stuff)
2. In `clone_default_group_in_public_schema`, replicate an event that (a) removes default access and (b) re-adds the roles. I think this should be able to reuse the old `RelationApiDualWriteGroupHandler` code and call `replicate_added_roles` with all the roles (or whatever based on above), similar to what `add_roles` used to do. I would probably add an argument or method to the group handler for removing default access, and then remove the `default_bindings_from_mapping` method from the bootstrap service. The logic there doesn't really have to be kept in the bootstrap service for a few reasons: 
  1. The group handler already knows that default access goes to the default workspace and already looks up the default workspace.
  2. It already has the group uuid and mapping for the default role binding uuid
  3. The only missing piece then is the platform default role uuid. We could consider just not removing this part since strictly speaking all we *have* to remove is the binding relationship. Regardless, the platform default role uuid is already repeated by the seeding process, so its something that should really be exposed outside of the bootstrap service in a reusable place, anyway.
3. Then, adjust `add_roles` and `remove_roles` only to use the modified methods per 1 above. (It would be an unrelated improvement to have a single outbox event for those–I hadn't realized or forgotten that was doing multiple I think.)

One easy way to improve the performance, too, would be to mint a `ReplicationEvent` and optionally allow passing _that_ into the add/remove roles methods. Then, if present, "merge" it with the one generated by _that_ method. This would decouple it from the logic and keep the "redundant" work purely in memory confined to a single CPU.

So in summary that would look something like this (psuedo-ish code):

```python
def clone_default_group_in_public_schema(group, tenant) -> Optional[Group]:
    """Clone the default group for a tenant into the public schema."""
    bootstrapped_tenant = None
    if settings.V2_TENANT_BOOTSTRAP:
        tenant_bootstrap_service = V2TenantBootstrapService(OutboxReplicator())
        bootstrapped_tenant = tenant_bootstrap_service.bootstrap_tenant(tenant)
        group_uuid = bootstrapped_tenant.mapping.default_group_uuid
    else:
        group_uuid = uuid4()

    public_tenant = Tenant.objects.get(tenant_name="public")
    # ... all the same group and policy set up stuff
    dual_write_handler = RelationApiDualWriteGroupHandler(group, ReplicationEventType.CUSTOMIZE_DEFAULT_GROUP)
    dual_write_handler.replicate_added_roles(role, remove_default_access_from=bootstrapped_tenant.mapping)
    return group
```

And then add/remove_roles are mostly untouched I think (aside from any updates to dual write handler interface).

WDYT?

Reply:
Two outbox events vs. one – it seems there are arguments for both approaches. From a performance standpoint, which is more efficient: processing one or two Kafka events (considering factors like serialization, deserialization, network, I/O, etc.)? Additionally, what it is preferable if one event in the sink fails: two separate events, or a single combined event? Also init of dual write handler twice.
Arguments which are  stated also valid. Personally I am not sure on 100 % what is optimal right now, so I choose your suggestion which (if I understand it well :) ) looks pretty simple and straightforward.

There are still some piece where clarification is needed:

> 1. Default binding tuples :
currently was described as tuples:

A: Workspace (default ws id) -> Role Binding ( from tenant mapping )
B: Role Binding (role binding from tenant mapping ) -> Role (Default policy UUID)
C: Role Binding (role binding from tenant mapping ) -> Default Group UUID (from tenant mapping)

you mentoined that not all of those are needed to add(create custom default role)/remove(delete custom default group) - is there the one which needs to be remove ?
I keept A and C - is also A necessary ?


> 2. One easy way to improve the performance, too, would be to mint a ReplicationEvent and optionally allow passing that into the add/remove roles methods. Then, if present, "merge" it with the one generated by that method. This would decouple it from the logic and keep the "redundant" work purely in memory confined to a single CPU.

This point is not clear to me -  what is meant by   
- "add/remove roles methods."  is it in dual write handler or in group/definer ?
- what is meant  "redundant" work here ?

I can see "redundant" work only in init dual write, otherwise now it calculates relations in memory and replicate two times as was suggested.

(Not sure if this is minor)

Reply:
A single event is definitely more efficient but kinda as I alluded to I think the cost is too great here. In general I think we should be more careful about performance, but mainly around unnecessary queries and especially around the read path. This is write path, and only a very particular write (modifying a default group for this first time).

For the tuples, we can just deal with adding and removing A. I added suggestions for this on the PR. That way we don't need any extra lookups (mainly to reduce the amount of code and logic) since we already have the mapping and the default workspace ID.

> This point is not clear to me - what is meant by
>
> "add/remove roles methods." is it in dual write handler or in group/definer ?

In the definer, so in this alteration you'd pass along the event as an argument (e.g. instead of the relationships that we had before.)

> what is meant "redundant" work here ?

Removing tuples that we just added – this could be worked out in memory rather than doing all the work in spicedb postgres to add and then remove the same tuples in a subsequent message. If it's all one we can just send one deduplicated batch to the sink.

> I can see "redundant" work only in init dual write, otherwise now it calculates relations in memory and replicate two times as was suggested.

What's redundant is that, in the worst case, you could be modifying the default group to remove all roles. In this case we'd publish two events: one event to add all tuples, and one event to remove them all. Better just not to add them in the first place. But it's not a big deal.

>
> (Not sure if this is minor)

I think it is definitely minor so I wouldn't worry about it.